shared-data/js/helpers/labwareSchemaShims.ts

```
@@ -64,6 +77,188 @@ export function getLabwareViewBox(
   }
 }
 
+/** Logically specifies where a labware is located. */
+export interface ComputeLabwareOriginInput {
+  /** The underlying deck. */
+  deckDefinition: DeckDefinition
+
+  /**
+   * The underlying slot. If this is a Flex and the labware is on a module, this should
+   * be the slot that the module replaces.
+   */
+  slotId: string
+
+  /**
+   * The definition of the module underneath the labware, if there is one.
+   * This should be specified even if it's only transitively underneath the labware,
+   * like slot -> module -> adapter -> labware in question.
+   */
+  moduleDefinition?: ModuleDefinition
+
+  /** The labware and all the labware underneath it. */
+  labwareDefinitionsBottomToTop: LabwareDefinition[]
+}
+
+/**
+ * Computes the absolute deck coordinates of a labware.
+ * Returns null if there was some internal error.
+ */
+export function computeLabwareOrigin(
+  input: ComputeLabwareOriginInput
+): Vector3D | null {
+  const stack = getLabwareStackAsArray(input)
+
+  if (stack == null) {
+    console.warn(
+      "Can't compute labware position. This is probably a bug in the call site, like passing in a nonexistent slot ID.",
+      input
+    )
+    return null
+  }
```

Is this a reasonable thing to do for error handling? If this were Python I would propagate an exception. But exceptions seem super unergonomic in TypeScript and we also don't use them much, so I'm sure it would cause whitescreens.

---

This is the type of thing I generally do, so I think it's reasonable :)  (for the exact reasoning you outline).